tools/odbc/connect/connect.cpp

```diff
@@ -0,0 +1,237 @@
+#include "connect.hpp"
+#include "duckdb/main/db_instance_cache.hpp"
+#include <iostream>
+
+#if WIN32
+#include <windows.h>
+#endif
+
+using namespace duckdb;
+
+//! The database instance cache, used so that multiple connections to the same file point to the same database object
+duckdb::DBInstanceCache instance_cache;
+
+bool Connect::SetSuccessWithInfo(SQLRETURN ret) {
+	if (SQL_SUCCEEDED(ret)) {
+		if (ret == SQL_SUCCESS_WITH_INFO) {
+			success_with_info = true;
+		}
+		return true;
+	}
+	return false;
+}
+
+static bool FindSubstrInSubstr(const std::string &s1, const std::string &s2) {
+	std::string longest = s1.length() >= s2.length() ? s1 : s2;
+	std::string shortest = s1.length() >= s2.length() ? s2 : s1;
+
+	idx_t longest_match = 0;
+	for (int i = 0; i < longest.length(); i++) {
+		for (int j = 0; j < shortest.length(); j++) {
+			if (longest[i] == shortest[j]) {
+				idx_t match = 1;
+				while (i + match < longest.length() && j + match < shortest.length() &&
+				       longest[i + match] == shortest[j + match]) {
+					match++;
+				}
+				if (match > longest_match) {
+					longest_match = match;
+				}
+			}
+		}
+	}
+
+	if (longest_match > 4) {
+		return true;
+	}
+	return false;
+}
+
+bool Connect::FindSimilar(const std::string &input, std::string &match) {
+	duckdb::vector<std::string> keys;
+	keys.reserve(conn_str_keynames.size());
+	for (auto &key_pair : conn_str_keynames) {
```

Comment:
You could use `StringUtil::CandidatesErrorMessage` for this, e.g.:

```cpp
StringUtil::CandidatesErrorMessage(conn_str_keynames, input, "Possible options", 5)
```

Reply:
Thanks, I was looking for something like this.